File: AlgorithmsPackJavaSolutions/Minimum Cost to Cut a Stick.java
class Solution {
    private int[][] memory;

    public int minCost(int n, int[] cuts) {
        int MAX = 100 + 1 + 2;
        memory = new int[MAX][MAX];
        for (int[] row : memory) {
            Arrays.fill(row, -1);
        }

        int[] cutsArray = new int[cuts.length + 2];
        System.arraycopy(cuts, 0, cutsArray, 0, cuts.length);
        cutsArray[cuts.length] = 0;
        cutsArray[cuts.length + 1] = n;
        Arrays.sort(cutsArray);

        return cost(0, cutsArray.length - 1, cutsArray);
    }

    private int cost(int start, int end, int[] cuts) {
        if (start + 1 == end) {
            return 0;
        }

        if (memory[start][end] != -1) {
            return memory[start][end];
        }

        int result = Integer.MAX_VALUE;
        for (int split = start + 1; split < end; split++) {
            int left = cost(start, split, cuts);
            int right = cost(split, end, cuts);
            int diff = cuts[end] - cuts[start];
            int total = left + right + diff;
            result = Math.min(result, total);
        }

        memory[start][end] = result;
        return result;
    }
}

File: AlgorithmsPackCPP/Minimum Cost to Cut a Stick.cpp
const int MAX = 100+1+2;
vector<int>cuts;
int memory[MAX][MAX];
class Solution {
public:
    int cost(int start, int end){
        if(start+1==end)
            return 0;
        
        auto&ret = memory[start][end];
        if(ret!=-1)
            return ret;
        
        ret = INT_MAX;
        
        for(int split = start+1; split<end; split++){
            int left = cost(start,split);
            int right = cost(split,end);
            int diff = cuts[end]-cuts[start];
            int total = left+right+diff;
            ret = min(ret,total);
        }
        return ret;
    }
    
    int minCost(int n, vector<int>& cuts_) {
        cuts = cuts_;
        cuts.push_back(0);
        cuts.push_back(n);
        sort(cuts.begin(),cuts.end());
        memset(memory,-1,sizeof(memory));
        return cost(0,cuts.size()-1);
    }
};
